### ISSUE TITLE
Add non-workspace specific foreground thread validator for MEF services

### ISSUE THREAD

**Issue by NTaylorMullen** [aspnet/Razor#1979]
This is a part of https://github.com/aspnet/Razor/issues/1973 but felt that it deserved its own issue since it's a key part of the Razor tooling system.

So the basic issue is that today we have a `ForegroundDispatcher` which is a workspace service; given that it's a workspace service this means that we need to have a workspace in order to get a foreground dispatcher. In the case of MEF services there are instances when we wont know what's the right workspace in VS for Mac (Such as the `RazorTextViewConnectionListener`), primarily because we'll be at a point when we haven't been given any text buffer or text view.

~Now you might be thinking ok, well lets break the connection between `ForegroundDispatcher` and workspaces; however, today our foreground dispatcher indirectly uses the `Workspaces` lifetime / instantiation threads to determine foreground/background thread task schedulers so we can dispatch between them. This is highly important. So we can't just make the dispatcher assume foreground/background schedulers (we need to rely on the workspaces lifetime to do that).~

~This issue tracks creating a `VisualStudioThreadValidator` (internal for now) that only has a `IsForegroundThread` property. Today we don't need to dispatch between foreground/background threads in the IDE's context.~

~I also want to mention in VS windows context the `VisualStudioThreadValidator` will functionally act exactly like the `ForegroundDispatcher.IsForegroundThread` since there's only a single workspace.~~

Edit: See below comment

**Comment by rynowak** [comment]
I'm interested to see if we can fix this without rewriting too much stuff. I really don't think we need new concepts.

If we make the `ForegroupDispatcher` a general-MEF thing and then add an extension method to get it given the workspacehostservices, that should solve this with very little churn.

**Comment by NTaylorMullen** [comment]
> I'm interested to see if we can fix this without rewriting too much stuff. I really don't think we need new concepts.
> 
> If we make the ForegroupDispatcher a general-MEF thing and then add an extension method to get it given the workspacehostservices, that should solve this with very little churn.

Actually, debugging through VS a bit more I realized that I made a bad assumption that our workspace service was constructed when the workspace was constructed. That means our foreground dispatcher today 100% depends on whoever is constructing it to be on the foreground thread. Indirectly that ends up being WTE. Given that finding I think we can make `ForegroundDispatcher` 100% MEF and have all of our workspace factories take that in (if the type they're constructing needs it).
